iris_grib/tests/unit/load_convert/test_product_definition_section.py

```diff
@@ -0,0 +1,134 @@
+            'productDefinitionTemplateNumber': template_number,
+            'parameterCategory': None,
+            'parameterNumber': None
+        })
+
+        fixed_surface_keys = [
+            'typeOfFirstFixedSurface',
+            'scaledValueOfFirstFixedSurface',
+            'typeOfSecondFixedSurface'
+        ]
+        for key in fixed_surface_keys:
+            # Force the presence or absence of the fixed surface elements even
+            # when they're respectively ignored or expected.
+            if fs_is_present and key not in section_4:
+                section_4[key] = template_0()[key]
+            elif (not fs_is_present) and key in section_4:
+                del section_4[key]
+
+        def run_function():
+            # For re-use in every type of test below.
+            product_definition_section(
+                section_4, metadata, discipline, tablesVersion, rt_coord)
+
+        if fs_is_expected and not fs_is_present:
+            # Should error since the expected keys are missing.
+            with self.assertRaises(KeyError) as context:
+                run_function()
+            self.assertTrue('FixedSurface' in str(context.exception))
+        else:
+            # Should have a successful run for all other circumstances.
+            run_function()
+            # Translate_phenomenon_patch is the end of the function,
+            # and should be able to accept None for the fixed surface
+            # arguments. So should always have run.
+            self.assertEqual(self.translate_phenomenon_patch.call_count, 1)
+            phenom_call_args = self.translate_phenomenon_patch.call_args[1]
+            for key in fixed_surface_keys:
+                # Check whether None or actual values have been passed for
+                # the fixed surface arguments.
+                if fs_is_expected:
+                    self.assertEqual(phenom_call_args[key], section_4[key])
+                else:
+                    self.assertEqual(phenom_call_args[key], None)
+
+    def test_fixed_surface_expected_present(self):
+        # Standard behaviour for most templates.
+        self.generic_fixed_surface_test(fs_is_expected=True,
+                                        fs_is_present=True)
+
+    def test_fixed_surface_ignored_absent(self):
+        # Standard behaviour for a few templates e.g. #31.
+        self.generic_fixed_surface_test(fs_is_expected=False,
+                                        fs_is_present=False)
+
+    def test_fixed_surface_expected_absent(self):
+        # Unplanned combination, should result in an error.
+        self.generic_fixed_surface_test(fs_is_expected=True,
+                                        fs_is_present=False)
+
+    def test_fixed_surface_ignored_present(self):
+        # Unplanned combination, should be handled same as 'ignored_absent'.
+        self.generic_fixed_surface_test(fs_is_expected=False,
+                                        fs_is_present=True)
```

Comment:
@trexfeathers Have you seen `itertools.product`? e.g.
```python
>>> from itertools import product
>>> [pairs for pairs in product([True, False], repeat=2]
[(True, True), (True, False), (False, True), (False, False)]
```
So the above four tests are simply expressed as:
```python
def test(self):
    '''Test all expected and present fixed-surface combinations.'''
    from itertools import product

    for pair in product([True, False], repeat=2):
        self.generic_fixed_surface_test(*pair)
```

Reply:
I have since encountered this in a home project. Wouldn't have occurred to me here, very neat.
_(Since this all runs in one test I have included a print within the function to show which combination is being tested. Also had to allow for a function call_count increasing rather than resetting)_